cbindgen/cbindgen.bzl

```diff
@@ -0,0 +1,220 @@
+            rust_lib.label,
+            rust_lib[CrateInfo].type,
+            supported_crate_types,
+        ))
+
+    # Determine the location of the cbindgen executable
+    if ctx.executable.cbindgen:
+        cbindgen_bin = ctx.executable.cbindgen
+    elif ctx.attr.cbindgen_path:
+        cbindgen_bin = ctx.executable.cbindgen_path
+    else:
+        cbindgen_bin = ctx.executable._cbindgen
+
+    # Optionally use the user defined template if one is provided
+    if ctx.file.config:
+        template_config = ctx.file.config
+        substitutions = ctx.attr.substitutions
+    else:
+        if ctx.attr.substitutions:
+            fail("'substitutions' ")
+
+        # Identify the desired language
+        use_c = ctx.attr.lang == "c"
+        use_cpp = ctx.attr.lang == "c++"
+
+        if use_c and use_cpp:
+            fail("The use of C/C++ must be mutually exclusive")
+
+        template_config = ctx.file._config_template
+        substitutions = {
+            "{language}": "C" if use_c else "C++",
+            "{include_guard}": "INCLUDE_{}_H".format(ctx.label.name.upper()),
+            "{namespace}": "namespace = \"{}\"".format(ctx.label.name) if use_cpp else "",
+        }
+
+    # Generate the `cbindgen.toml` config file
+    ctx.actions.expand_template(
+        template = template_config,
+        output = ctx.outputs.config,
+        substitutions = substitutions,
+    )
+
+    output_header = ctx.actions.declare_file(
+        ctx.attr.header_name if ctx.attr.header_name else "{}.h".format(ctx.label.name),
+    )
+
+    args = ctx.actions.args()
+    args.add("--config")
+    args.add(ctx.outputs.config)
+    args.add("--output")
+    args.add(output_header)
+    args.add(rust_lib[CrateInfo].root)
+
+    ctx.actions.run(
+        mnemonic = "RustCbindgen",
+        progress_message = "Generating {} C/C++ bindings for '{}'..".format(
+            ctx.attr.lang.upper(),
+            output_header.short_path,
+        ),
+        outputs = [output_header],
+        executable = cbindgen_bin,
+        inputs = rust_lib[CrateInfo].srcs + [ctx.outputs.config],
+        arguments = [args],
+    )
+
+    rust_compilation_context = rust_lib[CcInfo].compilation_context
+
+    # Add the new headers to the existing CompilationContext info
+    compilation_context = cc_common.create_compilation_context(
+        headers = depset([output_header], transitive = [rust_compilation_context.headers]),
+        defines = rust_compilation_context.defines,
+        framework_includes = rust_compilation_context.framework_includes,
+        includes = rust_compilation_context.includes,
+        local_defines = rust_compilation_context.local_defines,
+        quote_includes = rust_compilation_context.quote_includes,
+        system_includes = rust_compilation_context.system_includes,
+    )
+
+    # Return all providers given by `rust_library` to ensure compatiblity with other rules
```

Comment:
`rust_cbindgen_library` should effectively be transparent to other rules, in my opinion. It's job is only to inject a header into the outputs of a `rust_library`. So I thought the providers returned by the rule should match a `rust_library`

Reply:
I still think it would be weird for another rust_library to depend on this. 

It would make more sense if this was generated on-demand when a cc_library consumed a rust_library, but I have no idea how to do that. Maybe @damienmg knows something about that sort of thing?

Reply:
That actually sounds more ideal. All I want is to use a `rust_library` target with other targets that are compatible with `cc_library`. I do feel it's nicer to forward all the providers in this case though to avoid potential situations where people want the header file generated by this rule as well as other crate info. But yeah, I'd love to discuss a solution where `rust_library` just has a generated header.

Reply:
I actually think this is worth getting more input on @smklein @damienmg would you guys also be able to chime in on this? I think it would be fantastic if `rust_library` had a compatible header file such that it could be used in all places a `cc_library` was expected. I think this could simplify the implementation and be far more impactful and intuitive.